20240808-custom-limit-email-notifications.md

```diff
@@ -92,94 +107,131 @@ func ProjectNotificationFlagsFromInt(dbVal int) *ProjectNotificationFlags {
 func (p *ProjectNotificationFlags) Int() int {
 	toReturn := 0
 	if p.StorageExceeded80 {
-		toReturn |= storage80Filter
+		toReturn |= CustomStorage80
 	}
 	if p.StorageExceeded100 {
-		toReturn |= storage100Filter
+		toReturn |= CustomStorage100
 	}
     ...
 	return toReturn
 }
 ```
 
-If we use this approach, we only need to add a single updatable int column to the projects table, and since int has 32 bits, we would also be able to support any other similar flags we might want in the future, without schema changes.
+If we use this approach, we only need to add a single updatable int column to the projects table, and since int has 32 bits, we would also be able to support other similar threshold events in the future, without schema changes:
+
+```
+// project contains information about a user project.
+model project (
+    ...
+    // threshold_events is a bit string indicating whether certain limit-related events have occurred.
+    // Each event is associated with a different 
+    field threshold_events int ( updatable )
+```
+
+In addition to indicating which emails have been sent in the `projects` table, we also need to track project events in a queue, requiring a new table. The design and code for this table and its usage is similar to the existing `node_events` table, used for sending storage nodes emails based on reputation-related events:
+```
+// project_event table contains a project event queue for events which require an email notification.
+model project_event (
+	key id
+
+	index (
+		name project_events_id_created_at_index
```

Comment:
I thin this might need an index on created_at or last_attempted. That way the querying the events needing processing can be fast.

Reply:
Do you mean a separate index or added to this one?

Reply:
I guess in addition. I'm not quite sure which indexes will be necessary overall.